lib/modules/apostrophe-pages-headless/index.js

```diff
@@ -233,63 +236,153 @@ module.exports = {
-            }
-            var schema = manager.allowedSchema(req);
-            schema = self.addApplyToSubpagesToSchema(schema);
-            schema = self.removeParkedPropertiesFromSchema(existingPage, schema);
-            // overwrite fields that are in the schema
-            return self.apos.schemas.convert(req, schema, 'form', page, existingPage, callback);
-          },
-          update: function(callback) {
-            return self.update(req, existingPage, callback);
-          },
-          findAgain: function(callback) {
-            // Fetch the page. Yes, we already have it, but this way all the cursor
-            // filters run and we have access to ._url
-            return self.find(req, { _id: existingPage._id }).published(null).trash(self.apos.docs.trashInSchema ? null : false).toObject(function(err, _page) {
-              if (err) {
-                return callback(err);
-              }
-              if (!_page) {
-                return callback('notfound');
-              }
-              existingPage = _page;
-              return callback(null);
-            });
-          }
-        }, function(err) {
+
+        if (action === 'patch') {
+          return self.apos.locks.withLock('apostrophe-headless-' + id, body, respond);
+        } else {
+          return body(respond);
+        }
+
+        function respond(err) {
           if (err) {
             console.error(err);
             if (err === 'notfound') {
-              return res.status(404).send({ error: err });
+              return req.res.status(404).send({ error: err });
             } else if (err === 'invalid') {
-              return res.status(400).send({ error: err });
+              return req.res.status(400).send({ error: err });
             } else {
-              return res.status(500).send({ error: 'error' });
+              return req.res.status(500).send({ error: 'error' });
             }
           }
-          return res.send(existingPage);
-        });
-      });
+          return req.res.send(existingPage);
+        }
+
+        function body(callback) {
+          return async.series({
+            find: function(callback) {
+              return self.find(req, { _id: id }).permission('edit-apostrophe-page').trash(self.apos.docs.trashInSchema ? null : false).toObject(function(err, _page) {
+                if (err) {
+                  return callback(err);
+                }
+                if (!_page) {
+                  return callback('notfound');
+                }
+                existingPage = _page;
+                return callback(null);
+              });
+            },
+            convert: function(callback) {
+              var restApi = self.apos.modules['apostrophe-headless'];
+              var manager = self.apos.docs.getManager(self.apos.launder.string(page.type || existingPage.type));
+              if (!manager) {
+                // sneaky
+                return callback('notfound');
```

Comment:
Better to forbid a type that is legit but not a page type, however this is not a new issue.